### PR TITLE
Document Sub-Cameras

### PR DESCRIPTION
**Pull request by engineer124**
This takes the documentation of sub-cameras from https://github.com/zeldaret/oot/pull/908 and splits it into its own PR as more changes were needed to be addressed. This PR does the following:

- Consolidates the temp variables for overlays and code that use subCamera's.
- Tries to differentiate between Camera and Camera Id.
- For camera, use temps such as mainCam, activeCam, or subCam instead of Camera
- For camera Ids, use temps such as mainCamId, activeCamId, or subCamId
- Rename macros MAIN_CAM and all versions of SUB_CAM_ into CAM_ID_MAIN & CAM_ID_SUB_ to emphasize that they are camera Ids
- Remove CAM_ID_SUB_FREE 0 macro and replace it with CAM_ID_MAIN

For subCamera data inside actor struct, the following convention is used:
- `subCamId`, `subCamEye`, `subCamAt`, `subCamUp`
- `subCamEyeNext`, `subCamAtNext`
- `subCamEyeMaxVel` -> `subCamEyeMaxVelFrac` (Same for `At`)
- `subCamSpeedMod` -> `subCamVelFactor`

`Boss_Fd` and `Boss_Fd2` breaks apart from the rest by making it's own substruct to group all the `subCam` data together. I've left it as is as the style of how it was decomped.

Otherwise, I changed:

- In the `View` struct: `lookAt` -> `at` to match up with Camera
- In the `CutsceneContext` struct: `CameraPositions` -> `EyePoints` and `CameraFocus` -> `AtPoints` to also match camera terminology

### CODE REVIEW

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

File: src/overlays/actors/ovl_Boss_Ganon2/z_boss_ganon2.h [hunk 1]
```diff
@@ -79,11 +79,11 @@ typedef struct BossGanon2 {
...
     /* 0x039C */ s16 unk_39C;
-    /* 0x039E */ s16 unk_39E;
+    /* 0x039E */ s16 unk_39E; // subCamId
     /* 0x03A0 */ char unk_3A0[0x4];
-    /* 0x03A4 */ Vec3f unk_3A4;
-    /* 0x03B0 */ Vec3f unk_3B0;
-    /* 0x03BC */ Vec3f unk_3BC;
+    /* 0x03A4 */ Vec3f unk_3A4; // subCamEye
+    /* 0x03B0 */ Vec3f unk_3B0; // subCamAt
+    /* 0x03BC */ Vec3f unk_3BC; // subCamUp
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/overlays/actors/ovl_Boss_Ganon2/z_boss_ganon2.h:86]: These unknowns are involved in the Ganon2 `NON_MATCHING`, and my preference is to leave them as unknowns so it's marginally easier to read the asm with the code, although with the new line-numbers that's less of an issue

File: src/code/z_camera.c [hunk 2]
```diff
@@ -4951,7 +4951,7 @@ s32 Camera_Unique7(Camera* camera) {
 
     OLib_Vec3fDiffToVecSphGeo(&playerPosEyeOffset, eye, &playerPosRot->pos);
 
-    // fov actually goes unused since it's hard set later on.
+    // fov actually goes unused since its hard set later on.
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_camera.c:4954]: `change the camera to it's parent`, that was the `it's` I was trying to fix lol, somehow got mixed up

File: src/overlays/actors/ovl_En_Zl4/z_en_zl4.c [hunk 3]
```diff
@@ -180,24 +180,24 @@ static struct_80034EC0_Entry sAnimationEntries[] = {
...
-void EnZl4_SetCsCameraAngle(GlobalContext* globalCtx, s16 index) {
+void EnZl4_SetActiveCamDir(GlobalContext* globalCtx, s16 index) {
     Camera* activeCam = GET_ACTIVE_CAM(globalCtx);
 
     Camera_ChangeSetting(activeCam, CAM_SET_FREE0);
-    activeCam->at = sCsCameraAngle[index].at;
-    activeCam->eye = activeCam->eyeNext = sCsCameraAngle[index].eye;
-    activeCam->roll = sCsCameraAngle[index].roll;
-    activeCam->fov = sCsCameraAngle[index].fov;
+    activeCam->at = sActiveCamDir[index].at;
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Zl4/z_en_zl4.c:187]: I don't think `sActiveCamDir`/`sActiveCamMove` should mention the target camera being the active one, since it's not relevant to the data
Until a further documentation pass maybe gives it an actual name maybe just `sCamDirections`

File: src/overlays/actors/ovl_En_Heishi2/z_en_heishi2.h [hunk 4]
```diff
@@ -19,11 +19,11 @@ typedef struct EnHeishi2 {
     /* 0x0266 */ char unk_266[0x06];
     /* 0x026C */ Vec3s unk_26C; // padding inbetween these
     /* 0x0274 */ Vec3f unk_274; 
-    /* 0x0280 */ Vec3f unk_280; // camera related
-    /* 0x028C */ Vec3f unk_28C; // camera related
-    /* 0x0298 */ Vec3f unk_298; // camera related
+    /* 0x0280 */ Vec3f subCamEye;
+    /* 0x028C */ Vec3f subCamAt;
+    /* 0x0298 */ Vec3f subCamAtInit;
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Heishi2/z_en_heishi2.h:24]: Iirc some struct members like `subCamAtInit` `subCamEyeInit` are called `...Copy` in another place

Comment (engineer124) [reply src/overlays/actors/ovl_En_Heishi2/z_en_heishi2.h:24]: Yeah I realized that too very recently, they should be changed to `Init` now

File: src/overlays/actors/ovl_En_Zl4/z_en_zl4_cutscene_data.c [hunk 5]
```diff
@@ -178,7 +178,7 @@ static CutsceneCameraPoint D_80B5FA10[] = {
     { -1, 0, 0, 20.799965f, { -549, 124, 29 } }, { -1, 0, 0, 20.799965f, { -549, 124, 29 } },
 };
 
-static CutsceneCameraMove sCsCameraMove[] = {
+static CutsceneCameraMove sActiveCamMove[] = {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Zl4/z_en_zl4_cutscene_data.c:181]:
```suggestion
static CutsceneCameraMove sCamMove[] = {
```

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

**Review by Dragorn421 — APPROVED** [review]

**Comment by engineer124** [comment]
A few small extra changes were made to this PR:

 - In the context of cutscenes, `lookAt` is used in preference of `At` to be consistent with future cutscene macro changes.
 - Cutscene macros will be saved for a separate PR due to the number of cmds used in the repo.
 - Tiny cleanup that was previously missed

**Review by fig02 — APPROVED** [review]

**Comment by fig02** [comment]
there are a good number of conflicts that need to be fixed, but the changes in general look good to me

**Review by Dragorn421 — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c [hunk 6]
```diff
@@ -188,33 +188,33 @@ void func_808A3D58(BgMoriHineri* this, GlobalContext* globalCtx) {
...
-    if (globalCtx->activeCamera == sNextCamIdx) {
-        if (sNextCamIdx != MAIN_CAM) {
+    if (globalCtx->activeCamId == sSubCamId) {
+        if (sSubCamId != CAM_ID_MAIN) {
             objBankIndex = this->dyna.actor.objBankIndex;
             this->dyna.actor.objBankIndex = this->moriHineriObjIdx;
             this->moriHineriObjIdx = objBankIndex;
             this->dyna.actor.params ^= 1;
-            sNextCamIdx = MAIN_CAM;
+            sSubCamId = CAM_ID_MAIN;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c:209]: I see this type of change in a bunch of other actors but does it make sense for a sub camera ID to be set or compared to `CAM_ID_MAIN`? I feel like this would indicate they could be any camera and not necessarily a sub camera (at least as far as the actor code is concerned)

This makes me wonder if it's even right to be naming main and sub camera IDs differently in most cases, since the difference is already represented by the ID itself and clearly used as such in a lot of places like here

**Review by Roman971 — COMMENTED** [review]
Just a few things but looks good to me overall

File: src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c (hunk 6 above)
Comment (engineer124) [reply src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c:209]: This is an important point and part of the discussion as to why the `SUB_CAM_NONE = 0` define was introduced previously, but also later removed.

Basically, for all the variables that store `subCamId`, when it's set to 0, it is no longer used and becomes inactive. If a camera want's to access the main camera, it will always use `CAM_ID_MAIN` directly instead of going through the `subCamId` temps. So `subCamId` is never actually used for the main cam, it's always used through a temp. It is this idea of inactivity which was the reason `SUB_CAM_NONE` existed in the first place.

Likewise, when a subcamera is requested, it is checked which of the 3 subcamera's are available and assigns that to `subCamId`. It's not always the same subCam that get's assigned, basically it's the first one that's available (that way the game can handle multiple cutscenes at once). It seems distinct enough from the main camera to warrant it's own name to me imo, but there is a bit of a gray area here.

File: src/code/z_demo.c [hunk 7]
```diff
@@ -46,7 +46,7 @@ u32 D_8011E304[] = { 0x0200B650, 0x02015600, 0x02014F80, 0x02003F80, 0x02000330,
 u16 D_8015FCC0;
 u16 D_8015FCC2;
 u16 D_8015FCC4;
-s16 D_8015FCC6;
+s16 sActiveCamId;
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_demo.c:49]: Yeah true. Thoughts on `sPrevCamId`?

File: include/z64.h [hunk 8]
```diff
@@ -273,12 +273,12 @@ typedef struct {
...
     /* 0x12 */ u16   unk_12;
-    /* 0x14 */ s32   unk_14;
+    /* 0x14 */ s32   subCamId;
     /* 0x18 */ u16   unk_18;
     /* 0x1A */ u8    unk_1A;
     /* 0x1B */ u8    unk_1B;
-    /* 0x1C */ CutsceneCameraPoint* cameraFocus;
-    /* 0x20 */ CutsceneCameraPoint* cameraPosition;
+    /* 0x1C */ CutsceneCameraPoint* subCamLookAtPoints;
+    /* 0x20 */ CutsceneCameraPoint* subCamEyePoints;
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64.h:281]: It would always be for sub cameras yeah, I guess in the context of cutscenes it'd be fine without the sub

File: src/overlays/actors/ovl_Boss_Ganon2/z_boss_ganon2.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply src/overlays/actors/ovl_Boss_Ganon2/z_boss_ganon2.h]: These have been named now that Ganon2 is matching

File: src/code/z_demo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (Roman971) [reply src/code/z_demo.c:99]: Maybe something indicating it's the "gameplay" camera at that point, like `sGameplayCamId`, as opposed to the camera(s) used for cutscene purposes in this context? I think this could make the most sense since the system is also responsible for bringing the camera of this id back to be active once the cutscene is finished. Also it could fit with the `Gameplay_CopyCamera` call used at the end of some cutscenes (see `func_80068DC0`), which I'm guessing is meant to set the gameplay camera to the cutscene one at the end for a smoother transition.

Otherwise we could go with a more generic approach like "prev" or "save", but it might not be very clear, so I'm not sure.

File: src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c (hunk 6 above)
Comment (Roman971) [reply src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c:209]: Is there a real distinction between main and sub camera ids as far as the camera system is concerned? I feel like these `subCamId` variables could also be called `camId` and it would still make sense (`camId` being the id of the separate camera used in the context of the actor for instance), but I might be missing some internal details.

Otherwise, based on your explanation, I don't think anything treated as a "sub camera id" should be compared or set to `CAM_ID_MAIN`, so I guess we should at least use 0 or a dedicated define like the `SUB_CAM_NONE` you mentioned.

Basically I think we should either treat a camera id as a single thing and have no difference between sub and main camera ids, or completely separate them with different values/defines.

File: include/z64.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dragorn421) [reply include/z64.h:278]: The same point could be made about any "sub"-named member, I think on the contrary that using only sub-cameras means the prefix should stay, at least for consistency

File: src/code/z_demo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dragorn421) [reply src/code/z_demo.c:99]: I agree "active" is misleading
`sPrevCamId` sounds fine
or maybe `sPrevActiveCamId` / `sReturnToCamId` if you like that

File: src/overlays/actors/ovl_En_Ta/z_en_ta.c [hunk 9]
```diff
@@ -519,28 +519,28 @@ s32 func_80B14DD8(void) {
...
+    Vec3f subCamEye;
+    Vec3f subCamAt;
 
-    this->unk_2D0 = Gameplay_CreateSubCamera(globalCtx);
-    this->unk_2D2 = globalCtx->activeCamera;
-    Gameplay_ChangeCameraStatus(globalCtx, this->unk_2D2, CAM_STAT_WAIT);
-    Gameplay_ChangeCameraStatus(globalCtx, this->unk_2D0, CAM_STAT_ACTIVE);
+    this->subCamId = Gameplay_CreateSubCamera(globalCtx);
+    this->activeCamId = globalCtx->activeCamId;
+    Gameplay_ChangeCameraStatus(globalCtx, this->activeCamId, CAM_STAT_WAIT);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Ta/z_en_ta.c:527]: another spot where "active" is confusing, that's the previously active camera

**Comment by fig02** [comment]
Been a while since I've looked through this, are we still looking to resolve some stuff about what "active" means?

File: src/code/z_demo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_demo.c:99]: Apologies for the 4-month delay in responding, I haven't had the motivation for this particular PR but I'm getting back into it now :)

I personally like `sReturnToCamId` the most out of the suggestions, so I'll change it to that for now. But I'll keep this thread open for further discussion.

I'm not sure if I like the term "Gameplay" as it's not really used anywhere else to distinguish between the 4 camera's, and `Gameplay_CopyCamera` just leads to `Camera_Copy` anyway, and it just an extra layer to check for active camera id's.

File: src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c (hunk 6 above)
Comment (engineer124) [reply src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c:208]: Bump on this discussion as this was the main sticking point that I was split 50/50 on, and was a main factor in my delayed response. We can move this discussion to discord and I can copy the message below for reference. But for now:

The main point I was stuck on was whether to consider the 4 cameras as either:
- "4 identical camera" with the main vs. sub distinction only being in the application of the cameras
- "1 main camera + 3 sub cameras" with making a stronger differentiation between these.

Points for "4 identical camera"
- All 4 camera's have the same struct
- Main vs. sub is more of how it's used vs the struct itself

Points for "1 main camera + 3 sub camera"
- The "main camera" is clearly designed for normal game-play as link, while the 3 sub-cameras are clearly designed for cutscenes of various sorts, you can look at all the actors from this PR to see cases of this
- There are cases, particularly in "z_play", that treat the main cam different from the 3 sub-cams, particularly: https://github.com/zeldaret/oot/blob/32e66c2da8c2f9816773e1b2a0bc92804e9d20de/src/code/z_play.c#L218-L225 and the function `Gameplay_CreateSubCamera` https://github.com/zeldaret/oot/blob/32e66c2da8c2f9816773e1b2a0bc92804e9d20de/src/code/z_play.c#L1505-L1512

File: include/z64.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64.h:281]: I'll leave this open for now, but this ties into the whole "main vs. sub" discussion, and I'll address this once that's resolved

File: src/overlays/actors/ovl_En_Ta/z_en_ta.c (hunk 9 above)
Comment (engineer124) [reply src/overlays/actors/ovl_En_Ta/z_en_ta.c:527]: Would `this->returnToCamId` work in this case? This would apply to `en_ta`, `en_kz`, and `en_in` as they all have this same pattern and current name

Comment (Dragorn421) [reply src/overlays/actors/ovl_En_Ta/z_en_ta.c:527]: Sure! can discuss it further in the other thread, but it's ok to me. as long as the name isn't misleading (like `returnToCamId` is not) and is consistently used

**Review by Dragorn421 — APPROVED** [review]
Good for me once we get the debate settled on the camera indices defines (i.e. using `CAM_ID_MAIN` vs a hypothetical `CAM_ID_SUB_NONE`)

File: src/code/z_demo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (Roman971) [reply src/code/z_demo.c]: Yeah returnToCam seems fine as well. It's unusual with the rest of the codebase right now, but it should work

File: include/z64camera.h [hunk 10]
```diff
@@ -11,15 +11,14 @@
...
 #define NUM_CAMS 4
-#define MAIN_CAM 0
-#define SUBCAM_FIRST 1
-#define SUBCAM_FREE 0
-#define SUBCAM_NONE -1
-#define SUBCAM_ACTIVE -1
+#define CAM_ID_MAIN 0
+#define CAM_ID_SUB_FIRST 1
+#define CAM_ID_NONE -1
+#define CAM_ID_ACTIVE -1
```
Comment (Roman971) [inline include/z64camera.h:17]: I didn't notice this in my first review, but I feel like these defines should really be the same thing. From what I can tell, `CAM_ID_ACTIVE` is always interpreted as "use active cam", but I think we could also see this as a default behavior when running into `CAM_ID_NONE`, and just use that define all the time.

Also, it would be okay to have defines sharing the same value if they were used strictly during different scenarios, but it's not really the case here. In fact, there are a lot of sequences like `Gameplay_CreateSubCamera`, which may return -1, followed directly by `Gameplay_ChangeCameraStatus`, which has specific behavior based on -1, so using different defines doesn't really fit imo.

File: src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c (hunk 6 above)
Comment (Roman971) [reply src/overlays/actors/ovl_Bg_Mori_Hineri/z_bg_mori_hineri.c:208]: Going back to this discussion, I think it makes sense to treat main and sub cameras separately in the global context code (ie. the `Gameplay_` camera functions), because they clearly separated them on a lower level like you mentioned.

But when it comes to the rest of the codebase, I think it *could* make more sense to omit the distinction in the names, and rely on the camera ID to make the distinction between main camera (ID 0) and sub cameras (ID 1+).

However, if we do want to distinguish main and sub cameras everywhere, then I still think we should have a separate `SUB_CAM_ID_NONE` for this purpose (or maybe `SUB_CAM_ID_STOP` or `SUB_CAM_ID_DONE` which would avoid confusion with `CAM_ID_NONE`). Setting a sub cam id to main "works" if you already know what it means, but it still feels very wrong to me and would be very confusing for someone not familiar with the pattern.

I'm personally okay with either options, and I'm not sure which one would be better since both standpoints can be argued, but I really think we need to make a change in some direction

**Comment by Dragorn421** [comment]
## Summary of camera ids

Camera ids are indices into `cameraPtrs`
The first index is the main camera, the following are for "subcameras"
The cameras are parented to each other, parents being switched to when child cameras are removed (afaict).

The main camera is really set aside compared to others, it is defaulted to in many cases where stuff needs a camera but there's no explicit id to use. It is also stored on its own as `Camera mainCamera;` in the play state (sub cameras are `Camera subCameras[NUM_CAMS - CAM_ID_SUB_FIRST];`)

Some camera id values have special meaning, like `-1` in some contexts meaning "use active camera". We have defines for the special camera ids, but the weird thing is some constants (like `-1`) have different meaning in different contexts.
The question is, do we use different defines then, one per meaning, or only one per value?

From `include/z64camera.h` on master and in this PR:

| value | zeldaret/master | engineer124/subCams | usage |
| --- | --- | --- | --- |
| 4 | `NUM_CAMS` | `NUM_CAMS` | `Camera* cameraPtrs[NUM_CAMS];` |
| 0 | `MAIN_CAM` | `CAM_ID_MAIN` | `Gameplay_GetCamera(camera->globalCtx, CAM_ID_MAIN)`, `while (parentCam->childCamId != CAM_ID_MAIN)`, `if (sSubCamId != CAM_ID_MAIN)`  |
| 1 | `SUBCAM_FIRST` | `CAM_ID_SUB_FIRST` | `for (i = CAM_ID_SUB_FIRST; i < NUM_CAMS; i++)` |
| 0 | `SUBCAM_FREE` | currently -> `CAM_ID_MAIN` | 0/`CAM_ID_MAIN`/"`SUB_CAM_ID_NONE`" is used by actors when they don't actually "use" (manually control) the camera |
| -1 | `SUBCAM_NONE` | `CAM_ID_NONE` | `osSyncPrintf(VT_COL(RED, WHITE) "camera control: error: fulled sub camera system area\n" VT_RST); return CAM_ID_NONE;`, `this->subCamId = CAM_ID_NONE;` |
| -1 | `SUBCAM_ACTIVE` | `CAM_ID_ACTIVE` | `s16 camIdx = (camId == CAM_ID_ACTIVE) ? globalCtx->activeCamId : camId;` |


----------

This was discussed back in August 2021, I have no idea what I was thinking back then.
Today I don't deeply care about what happens with merging/splitting `CAM_ID_MAIN` vs "`SUB_CAM_ID_NONE`", `CAM_ID_NONE` vs `CAM_ID_ACTIVE`. There are arguments for and against each, nothing decisive really

Still, I think having both "`SUB_CAM_ID_NONE`" and `CAM_ID_NONE` would not be a great idea because of possible confusion

Also the naming with these should probably keep `CAM_ID_` as a prefix, as this PR switches as many variables and names as it could find to `camId` in a harmonization effort. So if anything imo "`SUB_CAM_ID_NONE`" -> "`CAM_ID_SUB_NONE`"

I think I would be okay with merging `CAM_ID_NONE` and `CAM_ID_ACTIVE` if as Roman says they can both be seen as "no specific camera" (so either default to active or no camera at all) 🤔
but the NONE in "`CAM_ID_SUB_NONE`" wouldn't mean "active"...

File: include/z64camera.h [hunk 11]
```diff
@@ -15,15 +15,20 @@
...
+
+// Camera IDs are indices into `GlobalContext.cameraPtrs`
+// The main camera is at index 0
+#define CAM_ID_MAIN 0
+// The sub cameras start at index 1
+#define CAM_ID_SUB_FIRST 1
+// -1 is used to indicate "no camera"
+#define CAM_ID_NONE -1
+// 0 is also used for variables holding sub camera IDs to indicate "no camera"
+#define SUB_CAM_ID_DONE 0
```
Comment (fig02) [inline include/z64camera.h:27]:
```suggestion
// Camera IDs are indices into `GlobalContext.cameraPtrs`
#define CAM_ID_MAIN 0 // The main camera is at index 0
#define CAM_ID_SUB_FIRST 1 // The sub cameras start at index 1
#define CAM_ID_NONE -1 // -1 is used to indicate "no camera"

#define SUB_CAM_ID_DONE 0 // In some actors, 0 is also used for variables holding sub camera IDs to indicate "subcam is finished"
```
just looks neater this way imo. And I added "in some actors" to the last define, cause afaik this doesnt mean anything to camera/z_play and is just a local meaning to some actors

Comment (fig02) [reply include/z64camera.h:27]: also can probably afford to take the values themselves out of the comments. you can look at the definition to see what the number is

Comment (fig02) [reply include/z64camera.h:27]: for example
```
#define CAM_ID_MAIN 0 // The index of the main camera
#define CAM_ID_SUB_FIRST 1 // The index sub cameras start at
#define CAM_ID_NONE -1 // Used to indicate no camera. Can be used to default to the active camera in some scenarios

#define SUB_CAM_ID_DONE 0 // Used in some actors for variables holding sub camera IDs to indicate "subcam is finished"

**Comment by fig02** [comment]
do we want to add some clarifying comments to the usages of `SUB_CAM_ID_DONE` when used to default to active cam? rest of this looks fine to me otherwise

File: include/z64camera.h (hunk 11 above)
Comment (fig02) [reply include/z64camera.h]: I didnt mean to remove // Camera IDs are indices into `GlobalContext.cameraPtrs` at the top, it probably still helps give context for what these groups of defines are for

File: src/overlays/actors/ovl_Door_Warp1/z_door_warp1.c [hunk 12]
```diff
@@ -579,8 +579,8 @@ void func_80999FE4(DoorWarp1* this, GlobalContext* globalCtx) {
     if (Message_GetState(&globalCtx->msgCtx) == TEXT_STATE_NONE) {
         Audio_PlaySoundGeneral(NA_SE_EV_LINK_WARP, &this->actor.projectedPos, 4, &gSfxDefaultFreqAndVolScale,
                                &gSfxDefaultFreqAndVolScale, &gSfxDefaultReverb);
-        OnePointCutscene_Init(globalCtx, 0x25E9, 999, &this->actor, MAIN_CAM);
-        Gameplay_CopyCamera(globalCtx, -1, sRutoWarpSubCamId);
+        OnePointCutscene_Init(globalCtx, 0x25E9, 999, &this->actor, CAM_ID_MAIN);
+        Gameplay_CopyCamera(globalCtx, CAM_ID_NONE, sRutoWarpSubCamId);
         Gameplay_ChangeCameraStatus(globalCtx, sRutoWarpSubCamId, CAM_STAT_WAIT);
         this->rutoWarpState = WARP_BLUE_RUTO_STATE_WARPING;
```
Comment (fig02) [inline src/overlays/actors/ovl_Door_Warp1/z_door_warp1.c:585]: perhaps `// passing `CAM_ID_NONE` here forces it to default to the active camera`, and in the other actor that does this

Comment (engineer124) [reply src/overlays/actors/ovl_Door_Warp1/z_door_warp1.c:585]: Specifically for `Gameplay_CopyCamera` or for all instances of `CAM_ID_NONE`, or a middle ground where the intent was for it to be "Active"

Comment (fig02) [reply src/overlays/actors/ovl_Door_Warp1/z_door_warp1.c:585]: just the two that used to be CAM_ID_ACTIVE or whatever, before we reverted it
think those were the only two cases

**Review by fig02 — APPROVED** [review]

**Comment by fig02** [comment]
(resolved the conflicts)

**Review by Roman971 — APPROVED** [review]
